### ISSUE TITLE
MPI Crash when many patches empty

### ISSUE THREAD

**Issue by joezuntz** [rmjarvis/TreeCorr#123]
Hi Mike,

I see the following crash when running `ng.process` in TreeCorr under MPI in a case where most of patches have no objects in because one of the catalogs covers a smaller area than the other.  I've not checked yet whether the input catalog is actually right but thought I should report anyway.  Please do close if this should come under user error.

Cheers,
Joe

```
WARNING: Some patch numbers do not contain any objects!
The following patch numbers have no objects: {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39}
This may be a problem depending on your use case.
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/global/cscratch1/sd/zuntz/TXPipe/txpipe/__main__.py", line 5, in <module>
    PipelineStage.main()
  File "/usr/local/lib/python3.6/dist-packages/ceci/stage.py", line 301, in main
    stage.execute(args)
  File "/usr/local/lib/python3.6/dist-packages/ceci/stage.py", line 387, in execute
    stage.run()
  File "/global/cscratch1/sd/zuntz/TXPipe/txpipe/twopoint.py", line 88, in run
    result = self.call_treecorr(i, j, k)
  File "/global/cscratch1/sd/zuntz/TXPipe/txpipe/twopoint.py", line 365, in call_treecorr
    xx = self.calculate_shear_pos(i, j)
  File "/global/cscratch1/sd/zuntz/TXPipe/txpipe/twopoint.py", line 474, in calculate_shear_pos
    ng.process(cat_j, cat_i, comm=self.comm)
  File "/usr/local/lib/python3.6/dist-packages/treecorr/ngcorrelation.py", line 403, in process
    self.finalize(varg)
  File "/usr/local/lib/python3.6/dist-packages/treecorr/ngcorrelation.py", line 303, in finalize
    self.cov = self.estimate_cov(self.var_method)
  File "/usr/local/lib/python3.6/dist-packages/treecorr/binnedcorr2.py", line 841, in estimate_cov
    return estimate_multi_cov([self], method)
  File "/usr/local/lib/python3.6/dist-packages/treecorr/binnedcorr2.py", line 1130, in estimate_multi_cov
    return _cov_jackknife(corrs, func)
  File "/usr/local/lib/python3.6/dist-packages/treecorr/binnedcorr2.py", line 1202, in _cov_jackknife
    v,w = _make_cov_design_matrix(corrs, plist, func)
  File "/usr/local/lib/python3.6/dist-packages/treecorr/binnedcorr2.py", line 1049, in _make_cov_design_matrix
    c._calculate_xi_from_pairs(cpairs)
  File "/usr/local/lib/python3.6/dist-packages/treecorr/ngcorrelation.py", line 464, in _calculate_xi_from_pairs
    self._sum([self.results[ij] for ij in pairs])
  File "/usr/local/lib/python3.6/dist-packages/treecorr/ngcorrelation.py", line 354, in _sum
    np.sum([c.raw_xi for c in others], axis=0, out=self.raw_xi)
  File "<__array_function__ internals>", line 6, in sum
  File "/usr/local/lib/python3.6/dist-packages/numpy/core/fromnumeric.py", line 2182, in sum
    initial=initial, where=where)
  File "/usr/local/lib/python3.6/dist-packages/numpy/core/fromnumeric.py", line 90, in _wrapreduction
    return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
ValueError: output parameter for reduction operation add has too many dimensions
```

**Comment by rmjarvis** [comment]
No, that's probably a bug.  Probably a np.sum or something similar on an empty array, which I was expecting to be built up to have some non-zero shape.  I'll try to reproduce.

**Comment by rmjarvis** [comment]
BTW, a workaround in the meanwhile if this is actually what you wanted to do (i.e. if the small catalog isn't user error) is to not use patches for the catalog with the small footprint.  You don't have to use patches on every input catalog.  You only need at least one of them to have patches to use all the patch-based stuff.

**Comment by rmjarvis** [comment]
I wasn't quite able to reproduce this.  It looks like you have 4 patches with some lenses in it.  I only managed to get the error when there was only 1 patch with any lenses.  Then it had trouble with the jackknife realization that omitted that patch, since there was nothing to sum up.
TBH, I'm not sure what the right thing to do in this case is.  Clearly jackknife is inappropriate here.  So maybe still an exception, but one that explains what went wrong a little more clearly?

**Comment by rmjarvis** [comment]
How about this...  Rather than raise an exception, emit the following warning:
```
WARNING: A xi for calculating the jackknife covariance has no patch pairs.  This probably means these patch specifications are inappropriate for these data.
```
And then I just use all 0's for the xi in this case.  I think it's probably almost always the case that if this happens, you should use a different patch specification.  E.g. not using patches for the lenses in this case.  But this way, if someone has a valid use case where this nonetheless happens, then it doesn't bomb out.

**Comment by joezuntz** [comment]
Hi Mike,

I've been trying to work around this, but have found some other issues - not sure if they are associated or not, but it seemed likely.   I've put the stuff to recreate them at NERSC, here:
```
/global/cfs/cdirs/lsst/groups/WL/users/zuntz/treecorr-issue
```

with command lines to recreate the issues and the results here:

```
shifter --image joezuntz/txpipe python tc_error.py shot
# works


shifter --image joezuntz/txpipe python tc_error.py jackknife
# fails with error below, though we are using patches - looks like this won't work without
# finalize=False?
#
# Traceback (most recent call last):
#   File "tc_error.py", line 63, in <module>
#     ng.process(lens, source, comm=comm)
#   File "/usr/local/lib/python3.6/dist-packages/treecorr/ngcorrelation.py", line 403, in process
#     self.finalize(varg)
#   File "/usr/local/lib/python3.6/dist-packages/treecorr/ngcorrelation.py", line 303, in finalize
#     self.cov = self.estimate_cov(self.var_method)
#   File "/usr/local/lib/python3.6/dist-packages/treecorr/binnedcorr2.py", line 841, in estimate_cov
#     return estimate_multi_cov([self], method)
#   File "/usr/local/lib/python3.6/dist-packages/treecorr/binnedcorr2.py", line 1130, in estimate_multi_cov
#     return _cov_jackknife(corrs, func)
#   File "/usr/local/lib/python3.6/dist-packages/treecorr/binnedcorr2.py", line 1184, in _cov_jackknife
#     npatch, all_pairs = _get_patch_nums(corrs, 'jackknife')
#   File "/usr/local/lib/python3.6/dist-packages/treecorr/binnedcorr2.py", line 1164, in _get_patch_nums
#     raise ValueError("Using %s covariance requires using patches."%name)
# ValueError: Using jackknife covariance requires using patches.





shifter --image joezuntz/txpipe mpirun -n 2 python tc_error.py shot mpi
# fails with error below - not clear what it means
# Traceback (most recent call last):
#   File "tc_error.py", line 63, in <module>
#     ng.process(lens, source, comm=comm)
#   File "/usr/local/lib/python3.6/dist-packages/treecorr/ngcorrelation.py", line 398, in process
#     self._process_all_cross(cat1, cat2, metric, num_threads, comm, low_mem)
#   File "/usr/local/lib/python3.6/dist-packages/treecorr/binnedcorr2.py", line 775, in _process_all_cross
#     self += temp
#   File "/usr/local/lib/python3.6/dist-packages/treecorr/ngcorrelation.py", line 339, in __iadd__
#     self._set_metric(other.metric, other.coords)
#   File "/usr/local/lib/python3.6/dist-packages/treecorr/binnedcorr2.py", line 856, in _set_metric
#     coords, metric = parse_metric(metric, coords1, coords2)
#   File "/usr/local/lib/python3.6/dist-packages/treecorr/util.py", line 540, in parse_metric
#     raise ValueError("Invalid coords %s"%coords)
# ValueError: Invalid coords None
```
```

**Comment by rmjarvis** [comment]
Joe, this is definitely bad error reporting, and I should do something more graceful with this input, but the proximate problem here is user error, I'm afraid.

You forgot to specify `sep_units`, so all the patch pairs had no results in this theta range.  Adding `sep_units: 'arcmin'` makes it work.

**Comment by rmjarvis** [comment]
Ah, and I think I found a more graceful way to handle this.  So on the #123 branch, it works as is, but gives all zeros for the xi and varxi.  Which is correct for this input.  
The problem if you're curious is that normally I don't add the pair-wise results to the `results` dict if all the npairs == 0 for that run, to make things a little faster when doing jackknife, etc.  But I changed it to go ahead and add the results anyway if the catalogs all use the same patch number.  So now it realizes that patches were used, but just have zeros for everything.
